### DISCUSSION TITLE
Consider migrating away from Java/Swing!

### DISCUSSION THREAD

**Discussion by ghost** [NationalSecurityAgency/ghidra#4190] [General]
It is frustrating to work with the current user interface which is written using Java/Swing.
I can understand that at the point when development was initiated, Java/Swing might have been a good choice for cross-platform user interface development, but there are options available these days which are more feature rich while delivering improved responsiveness and performance than Java/Swing.

Would you please "consider" moving the user interface to [Qt](https://www.qt.io/)?
Qt is written in C++ (mostly) and delivers near native platform widget responsiveness than Java/Swing.

**Comment by dragonmacher** [comment]
Ghidra's usage of Swing is pervasive enough to effectively be inseparable.   I contend it would be much easier to rewrite Ghidra from scratch.

In the interest of discussion, perhaps you could help us understand Swing's deficiency for your needs.  I personally do not find Swing's performance to be an issue for a desktop application like Ghidra.   Further, our code design is such that Swing is typically not the performance bottleneck.

**Reply by ghost** [reply]
I thought the core of Ghidra was written in C++ with **only** the user interface layer being written in Java/Swing.

The most irritating issue with the current Java/Swing interface is the lag and unpredictable unresponsiveness of various widget elements, relative to that of native binary widget elements, typically those written in C++, hence my suggestion to go with [Qt](https://www.qt.io/).

**Reply by dragonmacher** [reply]
Almost all of  Ghidra is written in Java.  The Decompiler is the most notable part that is written in C++.   

If you find lag in UI responsiveness, it is likely due to the data model and how we use locks.   In my experience, the Swing code has not added any noticeable lag.

**Reply by astrelsky** [reply]
> I thought the core of Ghidra was written in C++ with **only** the user interface layer being written in Java/Swing.
> 
> The most irritating issue with the current Java/Swing interface is the lag and unpredictable unresponsiveness of various widget elements, relative to that of native binary widget elements, typically those written in C++, hence my suggestion to go with [Qt](https://www.qt.io/).

What tool do you see this issue with? The only one I cut see this with is the debugger but it is still in development and is to be expected. 


Also it may just be my opinion but Qt is great... if you like non stl compatibility containers and strings as well as the numerous other problems from being
written for c++98.
